### PR TITLE
8326541: [AArch64] ZGC C2 load barrier stub should consider the length of live registers when spilling registers

### PR DESCRIPTION
**Pull request by JoshuaZhuwj**
Currently ZGC C2 load barrier stub saves the whole live register regardless of what size of register is live on aarch64.
Considering the size of SVE register is an implementation-defined multiple of 128 bits, up to 2048 bits,
even the use of a floating point may cause the maximum 2048 bits stack occupied.
Hence I would like to introduce this change on aarch64: take the length of live registers into consideration in ZGC C2 load barrier stub.

In a floating point case on 2048 bits SVE machine, the following ZLoadBarrierStubC2 

```
  ......
  0x0000ffff684cfad8:   stp     x15, x18, [sp, #80]
  0x0000ffff684cfadc:   sub     sp, sp, #0x100
  0x0000ffff684cfae0:   str     z16, [sp]
  0x0000ffff684cfae4:   add     x1, x13, #0x10
  0x0000ffff684cfae8:   mov     x0, x16
 ;; 0xFFFF803F5414
  0x0000ffff684cfaec:   mov     x8, #0x5414                     // #21524
  0x0000ffff684cfaf0:   movk    x8, #0x803f, lsl #16
  0x0000ffff684cfaf4:   movk    x8, #0xffff, lsl #32
  0x0000ffff684cfaf8:   blr     x8
  0x0000ffff684cfafc:   mov     x16, x0
  0x0000ffff684cfb00:   ldr     z16, [sp]
  0x0000ffff684cfb04:   add     sp, sp, #0x100
  0x0000ffff684cfb08:   ptrue   p7.b
  0x0000ffff684cfb0c:   ldp     x4, x5, [sp, #16]
  ......
```

could be optimized into:

```
  ......  
  0x0000ffff684cfa50:   stp     x15, x18, [sp, #80]
  0x0000ffff684cfa54:   str     d16, [sp, #-16]!                   // extra 8 bytes to align 16 bytes in push_fp()
  0x0000ffff684cfa58:   add     x1, x13, #0x10
  0x0000ffff684cfa5c:   mov     x0, x16
 ;; 0xFFFF7FA942A8
  0x0000ffff684cfa60:   mov     x8, #0x42a8                     // #17064
  0x0000ffff684cfa64:   movk    x8, #0x7fa9, lsl #16
  0x0000ffff684cfa68:   movk    x8, #0xffff, lsl #32
  0x0000ffff684cfa6c:   blr     x8
  0x0000ffff684cfa70:   mov     x16, x0
  0x0000ffff684cfa74:   ldr     d16, [sp], #16
  0x0000ffff684cfa78:   ptrue   p7.b
  0x0000ffff684cfa7c:   ldp     x4, x5, [sp, #16]
  ......
```

Besides the above benefit, when we know what size of register is live,
we could remove the unnecessary caller save in ZGC C2 load barrier stub when we meet C-ABI SOE fp registers.

Passed jtreg with option "-XX:+UseZGC -XX:+ZGenerational" with no failures introduced.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326541](https://bugs.openjdk.org/browse/JDK-8326541): [AArch64] ZGC C2 load barrier stub should consider the length of live registers when spilling registers (**Enhancement** - P4)

### Reviewers
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17977/head:pull/17977` \
`$ git checkout pull/17977`

Update a local copy of the PR: \
`$ git checkout pull/17977` \
`$ git pull https://git.openjdk.org/jdk.git pull/17977/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17977`

View PR using the GUI difftool: \
`$ git pr show -t 17977`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17977.diff">https://git.openjdk.org/jdk/pull/17977.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17977#issuecomment-1960896016)

### CODE REVIEW

**Comment by stooart-mon** [comment]
That is a welcome change - in light of possibly very large registers, we don't want to save more than is necessary.
It looks OK, but have you run a specific test that demonstrates it working?

**Comment by stooart-mon** [comment]
Thanks, that  helps - I can see you're saving/restoring the correct register lengths. Would it be possible to generate a testcase to test that registers are being saved/restored correctly?

The following is a testcase that is an example of where this testing is done, although in this PR's case it isn't subroutines, but load/store barriers:

https://github.com/openjdk/jdk/commit/4cd318756d4a8de64d25fb6512ecba9a008edfa1#diff-949a4a2f889be36be47e9b02b6d6cd1247768953b95a024f649878bac721fa04

**Comment by JoshuaZhuwj** [comment]
> Thanks, that helps - I can see you're saving/restoring the correct register lengths. Would it be possible to generate a testcase to test that registers are being saved/restored correctly?

@stooart-mon I had previously thought about how to write a good test case, but I did not think of a good way at that time. Let me rethink how to handle this gracefully. Thanks :-)

**Comment by JoshuaZhuwj** [comment]
A jtreg test case is created for this commit.
Please let me know if any further comments.
Thanks.

File: test/hotspot/jtreg/gc/z/TestRegistersPushPopAtZGCLoadBarrierStub.java
Comment (stooart-mon) [inline test/hotspot/jtreg/gc/z/TestRegistersPushPopAtZGCLoadBarrierStub.java:291]: In the event of failure, would it be possible to print the erroneous output? The output from the subprocesses, being directly piped in, doesn't lend itself to easy debugging. At first I thought there might be an option that could alter OutputAnalyzers output, but sadly not.

Comment (JoshuaZhuwj) [reply test/hotspot/jtreg/gc/z/TestRegistersPushPopAtZGCLoadBarrierStub.java]: > In the event of failure, would it be possible to print the erroneous output? The output from the subprocesses, being directly piped in, doesn't lend itself to easy debugging. At first I thought there might be an option that could alter OutputAnalyzers output, but sadly not.

Done. Thanks for your comments.

**Comment by JoshuaZhuwj** [comment]
@stooart-mon Thanks for your review. Please let me know if you have any other comments.
@fisk I would appreciate it if you could share your comments on this change since it follows your previous work done for x86.

**Review by fisk — APPROVED** [review]
This looks good to me and seems to follow a similar design to what I did on x86_64 vectors. Thanks for doing this!

**Comment by JoshuaZhuwj** [comment]
Thanks a lot for the review! @fisk

**Comment by JoshuaZhuwj** [comment]
Waiting for another review.

**Comment by JoshuaZhuwj** [comment]
Thank you a lot for the reviews! @stooart-mon @fisk @robcasloz

**Pull request closed without merging**
